### PR TITLE
bpo-38328: Speed up the creation time of constant list literals.

### PR DESCRIPTION
**Pull request by brandtbucher**
This PR contains a small change to the peephole optimizer that converts sequences of:

`LOAD_CONST(a)`, `LOAD_CONST(b)`, ..., `BUILD_LIST(n)`

to

`LOAD_CONST((a, b, ...))`, `BUILD_LIST_UNPACK(1)`

The improvement quickly becomes significant for lists larger than a few items:

| Elements | Speedup |
|----------|---------|
| 5        | \~5%    |
| 10       | \~20%   |
| 15       | \~25%   |
| 20       | \~30%   |
| 25       | \~35%   |
| 30       | \~45%   |
| 35       | \~50%   |

This can be tested on any version of Python by comparing the performance of `[0, 1, 2, ...]` vs `[*(0, 1, 2, ...)]`. The common cases of empty and single-element lists are not affected by this change.

This is related to [bpo-33325](https://bugs.python.org/issue33325), but that was an invasive change for all collection literals that had an unknown affect on performance. I've limited this one to lists and kept it to a few lines in the peephole optimizer.

https://bugs.python.org/issue38328

### CODE REVIEW

File: Lib/test/test_sys.py [hunk 1]
```diff
@@ -1222,7 +1222,7 @@ def get_gen(): yield 1
         # list
         samples = [[], [1,2,3], ['1', '2', '3']]
         for sample in samples:
-            check(sample, vsize('Pn') + len(sample)*self.P)
+            check(list(sample), vsize('Pn') + len(sample)*self.P)
```
Comment (brandtbucher) [inline Lib/test/test_sys.py:1225]: This is a `sys.getsizeof` test. The 3-element lists are overallocated, so we resize them here.

**Comment by brandtbucher** [comment]
@serhiy-storchaka Since you proposed something similar a while back, do you mind taking a look at this when you have the chance?

File: Python/peephole.c [hunk 2]
```diff
@@ -321,16 +321,25 @@ PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
...
+                    break;
+                }
                 if (j > 0 && lastlc >= j) {
                     h = lastn_const_start(codestr, op_start, j);
                     if (ISBASICBLOCK(blocks, h, op_start)) {
                         h = fold_tuple_on_constants(codestr, codelen,
-                                                    h, i+1, consts, j);
+                                                    h, i+1-islist, consts, j);
+                        if (h >= 0 && islist) {
+                            h = copy_op_arg(codestr, i, BUILD_LIST_UNPACK, 1, i+1);
```
Comment (serhiy-storchaka) [inline Python/peephole.c]: `h` is not used. Remove `h = `.

It can also be replaced with
```c
codestr[i] = PACKOPARG(BUILD_LIST_UNPACK, 1);
```

**Comment by brandtbucher** [comment]
Thanks for the feedback @serhiy-storchaka. Tests are passing, ready for another review.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by brandtbucher** [comment]
Thank @serhiy-storchaka. Is anything else needed here?

**Comment by serhiy-storchaka** [comment]
Nothing particular. I just want to fix the issue with list overallocation first. Your PR is good, thank you.

**Pull request closed without merging**